### PR TITLE
fix: recover incomplete PGlite data dirs

### PR DESCRIPTION
**Pull request by ikenyal**
## Problem

Issue #98 describes a local dev failure mode where `tsx watch` or another restart interrupts PGlite during first-boot initialization. That can leave the local `.bb-data/<fullId>` PostgreSQL directory partially written. On the next boot, `new PGlite(dataDir)` may abort before the dev server reaches `listen()`, so the app stays unreachable until the directory is manually moved or deleted.

## Changes

- Detect incomplete local PGlite data directories before opening PGlite.
- Treat a directory as initialized only after `PG_VERSION`, `base`, `global`, and `global/pg_control` exist.
- Recover partial PGlite leaf directories even when initialization stopped before `PG_VERSION` was written.
- Avoid quarantining parent directories that contain already-initialized PGlite child data directories.
- Move incomplete data directories to a `.corrupt-<timestamp>-<pid>-<suffix>` sidecar instead of deleting them.
- Keep empty directories and already-initialized data directories untouched.
- Add regression tests for partial init directories, including `PG_VERSION`-only, empty `base`/`global`, and pre-`PG_VERSION` leaf states.
- Add a patch changeset for `@aws-blocks/bb-data`.

This is intentionally narrower than deleting any non-empty directory: it only targets leaf directories with PGlite/PostgreSQL markers such as `PG_VERSION`, `base`, `global`, `pg_wal`, or `postgresql.conf`, and it skips parent directories that contain initialized child data directories.

This addresses the PGliteEngine recovery path from #98. It does not attempt to change the `tsx watch` / process-group shutdown behavior that can interrupt first-boot initialization in the first place.

## Validation

Manual before/after reproduction of the corrupt-state symptom:

```text
# Before/failure mode: opening a partial PGlite data dir directly
partial dir contents: PG_VERSION only
result: RuntimeError: unreachable

# After/fixed path: opening the same partial dir through PGliteEngine
[PGliteEngine] Moved incomplete PGlite data directory from .../partial to .../partial.corrupt-...; created a fresh directory. Delete matching .corrupt-* directories when they are no longer needed.
{"rows":[{"id":"ok"}],"hasBase":true,"hasGlobal":true}
```

Automated checks:

```text
npm run build -w packages/bb-data
npm run test -w packages/bb-data
./node_modules/.bin/changeset status --since=origin/main
./node_modules/.bin/tsx scripts/verify-changeset-coverage.ts
git diff --check
git diff --cached --check
```

## Checklist

- [x] PR description included
- [x] Tests are changed or added
- [x] Relevant documentation is changed or added (changeset)

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 1]
```diff
@@ -53,6 +62,22 @@ function cleanStaleLock(dataDir: string): void {
   }
 }
 
+function hasInitializedPgliteDataDir(dataDir: string): boolean {
+  return PGLITE_REQUIRED_DATA_DIR_ENTRIES.every((entry) => existsSync(join(dataDir, entry)));
+}
+
+function recoverIncompletePgliteDataDir(dataDir: string): void {
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts:91]: `recoverIncompletePgliteDataDir()` decides to delete the entire directory from a loose heuristic. A single matching filename triggering an unconditional recursive delete is broader than the stated intent of recovering an interrupted first boot. We could tighten the trigger so deletion only fires when the directory clearly looks like a real-but-incomplete PGlite init rather than on any single marker. For example, require at least `PG_VERSION` present (the unambiguous PGlite signature) before recreating, or require two or more markers; alternatively rename the incomplete dir to a .corrupt-<timestamp> sidecar instead of `rmSync` so the data is recoverable, then log its location.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 2]
```diff
@@ -53,6 +62,22 @@ function cleanStaleLock(dataDir: string): void {
...
+  const entries = readdirSync(dataDir);
+  if (entries.length === 0 || hasInitializedPgliteDataDir(dataDir)) return;
+
+  const looksLikePgliteInit = entries.some((entry) => PGLITE_DATA_DIR_MARKERS.includes(entry));
+  if (!looksLikePgliteInit) return;
+
+  rmSync(dataDir, { recursive: true, force: true });
+  mkdirSync(dataDir, { recursive: true });
+  console.warn(`[PGliteEngine] Recreated incomplete PGlite data directory ${dataDir}`);
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts]: The only signal that user-visible local state was just deleted is a single `console.warn`. The rest of the engine uses `console.log/console.debug` for benign events, so a destructive recreate is not distinguished in severity, and the message does not say what was removed.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 3]
```diff
@@ -53,6 +62,22 @@ function cleanStaleLock(dataDir: string): void {
   }
 }
 
+function hasInitializedPgliteDataDir(dataDir: string): boolean {
+  return PGLITE_REQUIRED_DATA_DIR_ENTRIES.every((entry) => existsSync(join(dataDir, entry)));
+}
+
+function recoverIncompletePgliteDataDir(dataDir: string): void {
+  const entries = readdirSync(dataDir);
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts]: Should we wrap the `readdirSync` in a try/catch that returns early on failure matching the defensive posture of `cleanStaleLock` above it?

Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Done. I wrapped readdirSync in a try/catch and return early on failure, so this recovery step stays best-effort like cleanStaleLock. If the directory is still unusable, the normal PGlite initialization path will surface that error.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 1 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts:91]: Thanks, agreed. I tightened the recovery trigger to require PG_VERSION before treating a non-empty directory as an incomplete PGlite init, and changed the recovery path to move the existing directory to a dataDir.corrupt-<timestamp> sidecar instead of deleting it recursively. The regression test now also verifies that the partial PG_VERSION-only directory is retained in the sidecar while a fresh data dir is created.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 2 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Updated. The warning now includes both the original dataDir and the sidecar path, and the recovery now renames the incomplete directory instead of deleting it so the previous local state remains available for inspection.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 4]
```diff
@@ -53,6 +54,40 @@ function cleanStaleLock(dataDir: string): void {
   }
 }
 
+function hasInitializedPgliteDataDir(dataDir: string): boolean {
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts:68]: hasInitializedPgliteDataDir only checks that PG_VERSION, base, and global *exist*, not that they have content. PGlite unpacks the data dir as a tar via loadTar in a single depth-first pass, so the global/ directory inode gets created before the files inside it (pg_filenode.map, pg_authid, …) are written. If a boot is interrupted in that window this returns true, recovery skips the dir, and the next `new PGlite(dataDir)` falls over with an opaque Postgres/WASM error instead of self-healing — which is the exact failure mode this change is meant to prevent. Can we add one cheap content check, e.g. also require `global/pg_filenode.map`? It's the first file written into global/ and Postgres won't start without it.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 5]
```diff
@@ -53,6 +54,40 @@ function cleanStaleLock(dataDir: string): void {
...
+  try {
+    entries = readdirSync(dataDir);
+  } catch {
+    return;
+  }
+  if (entries.length === 0 || hasInitializedPgliteDataDir(dataDir)) return;
+
+  if (!entries.includes('PG_VERSION')) return;
+
+  const corruptDataDir = nextCorruptDataDir(dataDir);
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts:103]: These quarantined dirs pile up forever — every recovery renames to `.corrupt-<timestamp>` and nothing ever prunes them, so a repeated crash-on-boot loop keeps spawning full data-dir siblings. Could we at least log a one-liner telling the dev they can delete the `*.corrupt-*` dirs, or cap how many we keep? A note in DESIGN.md would do too.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 6]
```diff
@@ -53,6 +54,40 @@ function cleanStaleLock(dataDir: string): void {
...
+    suffix++;
+  }
+  return candidate;
+}
+
+function recoverIncompletePgliteDataDir(dataDir: string): void {
+  let entries: string[];
+  try {
+    entries = readdirSync(dataDir);
+  } catch {
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts]: This catch swallows everything, not just ENOENT. If the dir exists but we can't read it (EACCES/EPERM), we silently bail here and the real problem resurfaces later as a more confusing PGlite error. Worth narrowing to `err.code === 'ENOENT'` and letting other errors through.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 7]
```diff
@@ -53,6 +54,40 @@ function cleanStaleLock(dataDir: string): void {
...
 
+function hasInitializedPgliteDataDir(dataDir: string): boolean {
+  return PGLITE_REQUIRED_DATA_DIR_ENTRIES.every((entry) => existsSync(join(dataDir, entry)));
+}
+
+function nextCorruptDataDir(dataDir: string): string {
+  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
+  let candidate = `${dataDir}.corrupt-${timestamp}`;
+  let suffix = 1;
+  while (existsSync(candidate)) {
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts]: Minor: there's a check-then-act gap here — existsSync, then the caller renameSyncs into the candidate. Practically fine since this is single-process local dev, but the loop reads like it fully reserves the name when it doesn't. A short random/pid suffix would close the window (and let you drop the loop); otherwise a quick comment noting the single-process assumption would help.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 8]
```diff
@@ -53,6 +54,40 @@ function cleanStaleLock(dataDir: string): void {
...
+    return;
+  }
+  if (entries.length === 0 || hasInitializedPgliteDataDir(dataDir)) return;
+
+  if (!entries.includes('PG_VERSION')) return;
+
+  const corruptDataDir = nextCorruptDataDir(dataDir);
+  renameSync(dataDir, corruptDataDir);
+  mkdirSync(dataDir, { recursive: true });
+  console.warn(
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts]: This is a fully handled, self-healing event, so console.warn feels a bit loud — it'll surface as a warning on every clean recovery. console.log/info reads better here.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 9]
```diff
@@ -12,6 +12,7 @@ const PG_UNIQUE_VIOLATION = '23505';
 
 /** PostgreSQL error code class for connection exceptions. */
 const PG_CONNECTION_EXCEPTION_CLASS = '08';
+const PGLITE_REQUIRED_DATA_DIR_ENTRIES = ['PG_VERSION', 'base', 'global'];
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.ts]: Could you add a short comment on why these three entries mean "initialised" and that the order matters (PG_VERSION is written first by loadTar)? The whole recovery heuristic leans on that assumption and it'd be easy for someone to break it by accident.

File: packages/bb-data/src/engines/pglite-engine.test.ts [hunk 10]
```diff
@@ -175,3 +175,22 @@ test('constructor does not crash when parent directory is missing', async () =>
   const rows = await engine.query('SELECT * FROM t');
   assert.deepStrictEqual(rows, []);
 });
+
+test('constructor recovers an incomplete PGlite init directory', async () => {
+  const partial = join(TEST_DIR, 'partial-init');
+  mkdirSync(partial, { recursive: true });
+  writeFileSync(join(partial, 'PG_VERSION'), '16\n');
```
Comment (soberm) [inline packages/bb-data/src/engines/pglite-engine.test.ts:182]: Nice-to-have: this covers the "PG_VERSION only" case but not "PG_VERSION + empty base/global", which is the false-positive state hasInitializedPgliteDataDir would let through today. Adding that case would lock the gap down.

File: packages/bb-data/src/engines/pglite-engine.ts [hunk 11]
```diff
@@ -53,6 +54,40 @@ function cleanStaleLock(dataDir: string): void {
...
+function recoverIncompletePgliteDataDir(dataDir: string): void {
+  let entries: string[];
+  try {
+    entries = readdirSync(dataDir);
+  } catch {
+    return;
+  }
+  if (entries.length === 0 || hasInitializedPgliteDataDir(dataDir)) return;
+
+  if (!entries.includes('PG_VERSION')) return;
```
Comment (hfurkanbozkurt) [inline packages/bb-data/src/engines/pglite-engine.ts]: This guard does double duty: it correctly protects a parent dir (the CLI's `new PGliteEngine('.bb-data')` root has no top-level PG_VERSION), but it also skips a corrupt leaf with NO PG_VERSION at all — the original #98 mode (a non-empty `.bb-data/<id>` lacking PG_VERSION aborts `new PGlite()`). Combined with the existence-vs-content gap already noted on this PR (global/ created before its files), the only recovered state is "PG_VERSION present, base/global missing" — a narrow slice of the interruption window.

Suggestion: detect a partial leaf affirmatively instead of gating on PG_VERSION existence — recover when the dir is non-empty, not complete *by content* (require PG_VERSION AND global/pg_control, not just entry existence), and is itself a data dir (has postmaster.opts/postmaster.pid/base/global), while never
touching a dir whose children are complete data dirs. That catches the no-PG_VERSION leaf and the empty-global/ window without re-introducing the parent-wipe risk this guard was added for.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 7 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Done. The sidecar path now includes the process id and a random suffix, so it no longer relies on an existsSync loop that appears to reserve the destination name.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 4 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts:68]: Done. I tightened the initialized check to require global/pg_control in addition to PG_VERSION/base/global, so a directory with empty base/global no longer gets treated as complete. I used pg_control as the content sentinel since it is present in a successfully initialized PGlite/Postgres global directory and Postgres cannot start without it.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 5 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts:103]: Updated the recovery log to explicitly say that matching .corrupt-* directories can be deleted when they are no longer needed. I kept automatic pruning out of this patch so recovery remains non-destructive and does not guess which quarantined data dir is safe to remove.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 6 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Done. The readdirSync catch now returns only for ENOENT and rethrows other filesystem errors, so EACCES/EPERM should surface directly instead of being hidden behind a later PGlite failure.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 8 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Done. This is now console.log instead of console.warn since the event is self-healing after the incomplete directory has been moved aside.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 9 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Added a short comment above the initialized-dir check. It calls out that PG_VERSION is written before the directory tree and that global/pg_control is used to confirm that global contains PostgreSQL state, so the heuristic is less likely to be simplified back to existence-only checks.

File: packages/bb-data/src/engines/pglite-engine.ts (hunk 11 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.ts]: Updated along these lines. Recovery no longer gates solely on PG_VERSION; it treats a non-empty directory with PGlite leaf markers as recoverable when it is not complete by content, while skipping directories that contain initialized child data dirs. I also moved stale postmaster.pid cleanup after recovery so a pre-PG_VERSION leaf can still be recognized, and added a regression test for the pre-PG_VERSION case.

File: packages/bb-data/src/engines/pglite-engine.test.ts (hunk 10 above)
Comment (ikenyal) [reply packages/bb-data/src/engines/pglite-engine.test.ts:182]: Added that regression case. The test now seeds PG_VERSION plus empty base/global directories and verifies that PGliteEngine quarantines the partial dir, creates a fresh usable data dir, and writes global/pg_control in the recovered directory.

**Review by soberm — APPROVED** [review]

**Review by hfurkanbozkurt — APPROVED** [review]

**Review by sarayev — APPROVED** [review]
